packages/preferences-api/src/browser/preference-service.ts

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (C) 2017 Ericsson and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { JSONExt } from '@phosphor/coreutils';
+import { injectable } from 'inversify';
+import { Event, Emitter, DisposableCollection } from '@theia/core/lib/common';
+import { PreferenceService, PreferenceChange } from './preference-protocol';
+import { PreferenceProvider, NewPreferencesEvent } from './preference-protocol';
+
+@injectable()
+export class PreferenceServiceImpl implements PreferenceService {
+    protected prefCache: { [key: string]: any }
+        = {};
+
+    protected readonly toDispose = new DisposableCollection();
+    protected readonly onPreferenceChangedEmitter = new Emitter<PreferenceChange>();
+
+    protected readonly readyProviders: PreferenceProvider[] = [];
+
+    protected resolveReady: () => void;
+
+    readonly ready = new Promise<void>(resolve => {
+        this.resolveReady = resolve;
+    });
+
+    protected isReady: boolean = false;
+
+    constructor(protected readonly preferenceProviders: PreferenceProvider[]) {
+        this.preferenceProviders.forEach(preferenceProvider => {
+            preferenceProvider.setClient({
+                onNewPreferences: event => this.onNewPreferences(event)
+            });
+
+            this.toDispose.push(preferenceProvider);
+            preferenceProvider.ready.then(() => {
+                this.readyProviders.push(preferenceProvider);
+                /* Sorted so that providers with more priority i.e workspace > user override user settings if found */
+                this.readyProviders.sort((provider1, provider2) =>
+                    provider1.priority - provider2.priority
```

Comment:
why do you need priority? could be injection order used, i see that you are able to control it